dymos/transcriptions/pseudospectral/pseudospectral_base.py

```diff
@@ -56,21 +55,62 @@ def setup_states(self, phase):
         else:
             indep = om.IndepVarComp()
 
-            for name, options in phase.state_options.items():
-                if not options['solve_segments'] and not options['connected_initial']:
-                    indep.add_output(name='states:{0}'.format(name),
-                                     shape=(num_state_input_nodes, np.prod(options['shape'])),
-                                     units=options['units'])
-
         num_connected = len([s for (s, opts) in phase.state_options.items() if opts['connected_initial']])
         prom_inputs = ['initial_states:*'] if num_connected > 0 else None
         phase.add_subsystem('indep_states', indep, promotes_inputs=prom_inputs,
                             promotes_outputs=['*'])
 
+    def configure_states(self, phase):
+        grid_data = self.grid_data
+        num_state_input_nodes = grid_data.subset_num_nodes['state_input']
+        indep = phase.indep_states
+        time_units = phase.time_options['units']
+        ode = phase._get_subsystem(self._rhs_source)
+
         # add all the des-vars (either from the IndepVarComp or from the indep-var-like
         # outputs of the collocation comp)
         for name, options in phase.state_options.items():
+
+            rate_src = options['rate_source']
+
+            # Handle states that point to a control.
+            # This feels a little hackish
+            if rate_src in phase.control_options:
+                targets = phase.control_options[rate_src]['targets']
+                if targets is not _unspecified:
+                    rate_src = targets[0]
+
+            # Handle states that point to another state. States must be declared in the right
+            # order.
+            # This feels a little hackish
+            if rate_src in phase.state_options and options['shape'] in (_unspecified, None):
+                options['shape'] = phase.state_options[rate_src]['shape']
+
+            full_shape, units = get_target_metadata(ode, name=name,
```

Comment:
shouldn't this be `get_source_metadata`, since the rate_source is an output?  I'm a bit surprised this works, honestly.

Reply:
Well, it works because I modified it to look for io metadata in both the 'inputs' and 'outputs' vectors. I never noticed we had a `get_source_metadata`, but on first glance, it wouldn't work because it doesn't have the "targets" argument.  I suppose it could be modified to include that, but then there would be very little difference between the two functions.